src/backend/distributed/transaction/lock_graph.c

```diff
@@ -0,0 +1,752 @@
+{
+	char *resultString = NULL;
+	Datum resultStringDatum = 0;
+	Datum timestampDatum = 0;
+
+	if (PQgetisnull(result, rowIndex, colIndex))
+	{
+		return 0;
+	}
+
+	resultString = PQgetvalue(result, rowIndex, colIndex);
+	resultStringDatum = CStringGetDatum(resultString);
+	timestampDatum = DirectFunctionCall3(timestamptz_in, resultStringDatum, 0, -1);
+
+	return DatumGetTimestampTz(timestampDatum);
+}
+
+
+/*
+ * dump_local_wait_edges returns wait edges for distributed transactions
+ * running on the node on which it is called, which originate from the source node.
+ */
+Datum
+dump_local_wait_edges(PG_FUNCTION_ARGS)
+{
+	int32 sourceNodeId = PG_GETARG_INT32(0);
+
+	WaitGraph *waitGraph = BuildWaitGraphForSourceNode(sourceNodeId);
+	ReturnWaitGraph(waitGraph, fcinfo);
+
+	return (Datum) 0;
+}
+
+
+/*
+ * ReturnWaitGraph returns a wait graph for a set returning function.
+ */
+static void
+ReturnWaitGraph(WaitGraph *waitGraph, FunctionCallInfo fcinfo)
+{
+	ReturnSetInfo *rsinfo = (ReturnSetInfo *) fcinfo->resultinfo;
+	TupleDesc tupdesc = NULL;
+	Tuplestorestate *tupstore = NULL;
+	MemoryContext per_query_ctx = NULL;
+	MemoryContext oldcontext = NULL;
+	size_t curEdgeNum = 0;
+
+	/* check to see if caller supports us returning a tuplestore */
+	if (rsinfo == NULL || !IsA(rsinfo, ReturnSetInfo))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg(
+					 "set-valued function called in context that cannot accept a set")));
+	}
+	if (!(rsinfo->allowedModes & SFRM_Materialize))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("materialize mode required, but it is not " \
+						"allowed in this context")));
+	}
+
+	/* Build a tuple descriptor for our result type */
+	if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
+	{
+		elog(ERROR, "return type must be a row type");
+	}
+
+	per_query_ctx = rsinfo->econtext->ecxt_per_query_memory;
+	oldcontext = MemoryContextSwitchTo(per_query_ctx);
+
+	tupstore = tuplestore_begin_heap(true, false, work_mem);
+	rsinfo->returnMode = SFRM_Materialize;
+	rsinfo->setResult = tupstore;
+	rsinfo->setDesc = tupdesc;
+	MemoryContextSwitchTo(oldcontext);
+
+	/*
+	 * Columns:
+	 * 00: waiting_pid
+	 * 01: waiting_node_id
+	 * 02: waiting_transaction_num
+	 * 03: waiting_transaction_stamp
+	 * 04: blocking_pid
+	 * 05: blocking__node_id
+	 * 06: blocking_transaction_num
+	 * 07: blocking_transaction_stamp
+	 * 08: blocking_transaction_waiting
+	 */
+	for (curEdgeNum = 0; curEdgeNum < waitGraph->numUsed; curEdgeNum++)
+	{
+		Datum values[9];
+		bool nulls[9];
+		WaitEdge *curEdge = &waitGraph->edges[curEdgeNum];
+
+		memset(values, 0, sizeof(values));
+		memset(nulls, 0, sizeof(nulls));
+
+		values[0] = Int32GetDatum(curEdge->waitingPid);
+		values[1] = Int32GetDatum(curEdge->waitingNodeId);
+		if (curEdge->waitingTransactionNum != 0)
+		{
+			values[2] = Int64GetDatum(curEdge->waitingTransactionNum);
```

Comment:
This mostly determines how the user sees the output of `dump_*_wait_edges`. I think NULL is slightly prettier because it usually shows up as empty.